pkg/flowaggregator/flowlogger/logger_test.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2023 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package flowlogger
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	flowrecordtesting "antrea.io/antrea/pkg/flowaggregator/flowrecord/testing"
+)
+
+func getTestFlowLogger(maxLatency time.Duration) (*FlowLogger, *bytes.Buffer) {
+	var b bytes.Buffer
+	flowLogger := &FlowLogger{
+		maxLatency: maxLatency,
+		writer:     bufio.NewWriter(&b),
+	}
+	return flowLogger, &b
+}
+
+func TestWriteRecord(t *testing.T) {
+	record := flowrecordtesting.PrepareTestFlowRecord()
+
+	testCases := []struct {
+		prettyPrint bool
+		expected    string
+	}{
+		{
+			prettyPrint: true,
+			expected:    "1637706961,1637706973,10.10.0.79,10.10.0.80,44752,5201,TCP,perftest-a,antrea-test,k8s-node-control-plane,perftest-b,antrea-test-b,k8s-node-control-plane-b,10.10.1.10,5202,perftest,test-flow-aggregator-networkpolicy-ingress-allow,antrea-test-ns,test-flow-aggregator-networkpolicy-rule,Drop,K8sNetworkPolicy,test-flow-aggregator-networkpolicy-egress-allow,antrea-test-ns-e,test-flow-aggregator-networkpolicy-rule-e,Invalid,Invalid",
+		},
+		{
+			prettyPrint: false,
+			expected:    "1637706961,1637706973,10.10.0.79,10.10.0.80,44752,5201,6,perftest-a,antrea-test,k8s-node-control-plane,perftest-b,antrea-test-b,k8s-node-control-plane-b,10.10.1.10,5202,perftest,test-flow-aggregator-networkpolicy-ingress-allow,antrea-test-ns,test-flow-aggregator-networkpolicy-rule,2,1,test-flow-aggregator-networkpolicy-egress-allow,antrea-test-ns-e,test-flow-aggregator-networkpolicy-rule-e,5,4",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("pretty print: %t", tc.prettyPrint), func(t *testing.T) {
+			flowLogger, b := getTestFlowLogger(MaxLatency)
+			err := flowLogger.WriteRecord(record, tc.prettyPrint)
+			require.NoError(t, err)
+			flowLogger.Flush()
+			assert.Contains(t, b.String(), tc.expected)
+		})
+	}
+}
+
+func TestFlushLoop(t *testing.T) {
+	flowLogger, b := getTestFlowLogger(100 * time.Millisecond)
+	record := flowrecordtesting.PrepareTestFlowRecord()
+	err := flowLogger.WriteRecord(record, false)
+	require.NoError(t, err)
+	assert.Equal(t, 0, b.Len())
+	stopCh := make(chan struct{})
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		flowLogger.FlushLoop(stopCh)
+	}()
+	assert.Eventually(t, func() bool {
+		return b.Len() > 0
```

Comment:
It seems this line of code brings a race condition error https://github.com/antrea-io/antrea/actions/runs/4684072612/jobs/8299863482?pr=4855

Reply:
there is indeed a race, `bytes.Buffer` is not thread-safe